File: corr.ml
let correlate corr_size base mask sample_size =
  let correlation = Array.create (corr_size*corr_size) 0.0 in
  for oy = 0 to corr_size-1 do
    for ox = 0 to corr_size-1 do
      let sum = ref 0.0 in
      for y = 0 to sample_size-oy-1 do
        let base_idx = ((y+oy)*sample_size+ox) * 4 in
        let mask_idx = (y * sample_size * 4) in
        for x = 0 to sample_size-ox-1 do
          let b = base_idx+4*x and m = mask_idx+4*x in
          sum := !sum +. (
            base.(b) *. mask.(m) +.
            base.(b+1) *. mask.(m+1) +.
            base.(b+2) *. mask.(m+2) +.
            base.(b+3) *. mask.(m+3)
          )
        done
      done;
      correlation.(oy*corr_size+ox) <- !sum
    done
  done;
  correlation

let mksz sz =
  Array.init (sz*sz*4) (fun i -> float i *. 0.000001)

let () =
  let base = mksz 128 in
  let mask = mksz 128 in
  ignore (correlate 64 base mask 128)

File: corr.py
def correlate(corr_size, base, mask, sample_size):
  correlation = range(corr_size*corr_size)
  for offset_y in xrange(corr_size):
    for offset_x in xrange(corr_size):
      sum = 0
      for y in xrange(sample_size-offset_y):
        mask_idx = 4*(y*sample_size)
        base_idx = 4*((y+offset_y)*sample_size + offset_x)
        for x in xrange(sample_size-offset_x):
          sum += (
            base[base_idx] * mask[mask_idx] +
            base[base_idx+1] * mask[mask_idx+1] +
            base[base_idx+2] * mask[mask_idx+2] +
            base[base_idx+3] * mask[mask_idx+3] )
          mask_idx += 4
          base_idx += 4
      correlation[offset_y*corr_size + offset_x] = sum
  return correlation

def mklist(ssz):
  return [ i * 0.000001 for i in xrange(ssz*ssz*4) ]

base = mklist(64)
mask = mklist(64)
correlate(32, base, mask, 64)
